test/mlir/onnx/onnx_lowering_gru_op.mlir

```diff
@@ -16,23 +16,26 @@ func private @test_gru_forward_mode(%arg0: tensor<7x2x3xf32>, %arg1: tensor<1x12
 // CHECK-DAG:       [[VAR_c3_:%.+]] = arith.constant 3 : index
 // CHECK-DAG:       [[VAR_c2_:%.+]] = arith.constant 2 : index
 // CHECK-DAG:       [[VAR_c0_:%.+]] = arith.constant 0 : index
+// CHECK-DAG:       [[UCC_PARAM_3:%.+]] = builtin.unrealized_conversion_cast %arg3 : memref<1x24xf32> to tensor<1x24xf32>
```

Comment:
Just my curious to learn who generated this  `unrealized_conversion_cast`. Could you please elaborate a bit? thanks!

Reply:
The `unrealized_conversion_cast` is injected by MLIR after lowering the operations to reconcile types. It was introduced in MLIR around Jan 2021, this is the RFC https://llvm.discourse.group/t/open-meeting-1-14-dialect-conversion-and-type-conversion-the-question-of-cast-operations/2566. Essentially an `unrealized_conversion_cast` is a promise that the source and destination types will converge eventually. It is used to allow temporary intermixing of type systems during partial lowering.

Iy serve a similar purpose to our own `krnl.dummy_cast` AFAIK, and in this PR I also had to make the lowering of `onnx.Loop` and `onnx.Scan` use the new cast opcode rather than dummy casts.  I wanted to keep the "surface" of this PR as small as possible, so I plan to completely remove dummy_cast in an upcoming PR.